thriftbp/client_pool.go

```diff
@@ -291,9 +290,37 @@ type clientPool struct {
 
 	poolExhaustedCounter metrics.Counter
 	releaseErrorCounter  metrics.Counter
+
+	wrappedClient thrift.TClient
 }
 
 func (p *clientPool) Call(ctx context.Context, method string, args, result thrift.TStruct) (err error) {
+	// A clientPool needs to be set up properly before it can be used,
+	// specifically use p.wrapCalls to set up p.wrappedClient before using it.
+	//
+	// newClientPool already takes care of this for you.
+	return p.wrappedClient.Call(ctx, method, args, result)
```

Comment:
It _might_ make sense to be more defensive here and do a nil check on wrappedClient, fallback to use pooledCall directly if that's nil, but since this is an unexported type I don't feel strongly that's needed.

Reply:
Could also use a mutex or something to set it up lazily by just keeping track of the middleware?

Reply:
I'm ok with the fallback though, that seems reasonable.

Reply:
That's definitely overkill. the appropriate thing for lazy init is `sync.Once`, but if a few calls happen at the same time before it's initialized, they all need to wait for the initialize to finish. consider we usually initialize pools in main function shift the initialization delay there sounds more efficient.

The only thing defensive coding here prevents is someone not initialize the pool properly. Since it's unexported, that could only happen inside this package, which is fine.

Reply:
Ok, I think it's probably fine to leave this as-is?

Reply:
yes :)